pacstall/api/config_facade.py

```diff
@@ -0,0 +1,259 @@
+        parsed_url = __parse_bitbucket_url(url)
+
+    return parsed_url
+
+
+def __validate_attribute(
+    repo_name: str, conf: Dict[str, Optional[str]], attribute_name: str, ttype: type
+) -> bool:
+    """
+    Checks if the given `attribute_name` is part of `conf` and has the type `type`.
+
+    Returns
+    -------
+    If any of the conditions fail, an error message is printed and returns `False`, otherwise returns `True`.
+    """
+
+    if attribute_name not in conf:
+        fancy(
+            "error",
+            f"Repository '{repo_name}' is missing required attribute '{attribute_name}'",
+        )
+        return False
+    if type(conf[attribute_name]) is not ttype:
+        fancy(
+            "error",
+            f"Repository '{repo_name}' attribute '{attribute_name}' must be of type '{ttype}'",
+        )
+        return False
+    return True
+
+
+def __raise_unreachable() -> NoReturn:
+    """
+    Small hack to help the type checker in complex cases.
+    """
+    raise Exception("Unreachable code. This will never be raised.")
+
+
+def read_config() -> Tuple[Optional[ErrorCodes], Optional[List[RepositoryConfig]]]:
+    """
+    Reads and parses the repository list.
+
+    Returns
+    -------
+    A tuple consisting of the error code, and the repository list.
+    If the error code is `None` then the repository list will *not* be `None` and vice-versa.
+    """
+    try:
+        config_dict: Optional[Dict[str, Dict[str, Dict[str, Optional[str]]]]] = None
+        try:
+            with open(PACSTALL_CONFIG_PATH) as file:
+                config_dict = tomli.load(file)  # type: ignore[arg-type]
+        except OSError as error:
+            fancy(
+                "error",
+                f"Could not read repositories from file '{PACSTALL_CONFIG_PATH}'.\n{error}",
+            )
+
+            return (ErrorCodes.NO_INPUT_ERROR, None)
+
+        parsed_repo_list: List[RepositoryConfig] = []
+        for (repo_name, conf) in config_dict["repository"].items():
+            if not __validate_attribute(
+                repo_name, conf, attribute_name="url", ttype=str
+            ) or not __validate_attribute(
+                repo_name, conf, attribute_name="branch", ttype=str
+            ):
+                return (ErrorCodes.CONFIG_ERROR, None)
+
+            # Help type checker recognize `[conf["url"]` is not None
+            url = conf["url"] if conf["url"] is not None else __raise_unreachable()
+            parsed_url: Optional[str] = parse_url(url)
+            if parsed_url is None:
+                fancy(
+                    "error",
+                    f"Repository '{repo_name}' has invalid attribute 'url': {url}",
+                )
+                return (ErrorCodes.CONFIG_ERROR, None)
+
+            # Help type checker recognize `[conf["branch"]` is not None
+            branch = (
+                conf["branch"] if conf["branch"] is not None else __raise_unreachable()
+            )
```

Comment:
Why not just `# ignore:` this line instead.

Reply:
Didn't work for me

Reply:
nvm was using it work